File: src/main/mods/data/infrastructure/xbox-webapi-provider.ts
import XboxWebApiModule from 'xbox-webapi'
import type { DataSessionContext, XboxWebApiClient } from '../domain/types'
import { resolveWebTokenClaims } from '../domain/web-token'

type XboxWebApiCtor = new (config: { uhs: string; token: string }) => XboxWebApiClient

function resolveXboxWebApiCtor(): XboxWebApiCtor {
  const moduleValue = XboxWebApiModule as unknown as {
    default?: unknown
  }
  // 兼容 CJS/ESM 导出差异，统一拿到可构造的类
  const maybeCtor = (moduleValue.default ?? XboxWebApiModule) as unknown
  if (typeof maybeCtor !== 'function') {
    throw new TypeError('xbox-webapi constructor is unavailable')
  }
  return maybeCtor as XboxWebApiCtor
}

/**
 * xbox-webapi 客户端提供器
 * - 以 token 指纹维度维护单例，避免重复初始化
 */
export class XboxWebApiProvider {
  private currentClient: XboxWebApiClient | undefined
  private currentFingerprint = ''

  getOrCreate(session: DataSessionContext): XboxWebApiClient | undefined {
    const claims = resolveWebTokenClaims(session.webToken)
    if (claims === null) {
      return undefined
    }

    const fingerprint = `${claims.uhs}:${claims.userToken}`
    if (this.currentClient !== undefined && this.currentFingerprint === fingerprint) {
      return this.currentClient
    }

    const XboxWebApi = resolveXboxWebApiCtor()
    this.currentClient = new XboxWebApi({
      token: claims.userToken,
      uhs: claims.uhs
    })
    this.currentFingerprint = fingerprint
    return this.currentClient
  }

  getCurrent(): XboxWebApiClient | undefined {
    return this.currentClient
  }
}

File: src-tauri/src/mods/data/infrastructure/xbox_webapi_provider.rs
use crate::mods::data::client::XboxWebApiClient;
use crate::mods::data::domain::web_token::resolve_web_token_claims;
use crate::mods::data::domain::DataSessionContext;
use std::sync::Arc;

pub struct XboxWebApiProvider {
    current_client: Option<Arc<XboxWebApiClient>>,
    current_fingerprint: String,
}

impl XboxWebApiProvider {
    pub fn new() -> Self {
        Self {
            current_client: None,
            current_fingerprint: String::new(),
        }
    }

    // 按 token 指纹复用 client，避免重复初始化 HTTP 资源。
    pub fn get_or_create(&mut self, session: &DataSessionContext) -> Option<Arc<XboxWebApiClient>> {
        let claims = resolve_web_token_claims(&session.web_token)?;
        let fingerprint = format!("{}:{}", claims.uhs, claims.user_token);

        if let Some(client) = &self.current_client {
            if self.current_fingerprint == fingerprint {
                return Some(client.clone());
            }
        }

        let next_client = Arc::new(XboxWebApiClient::new(claims.uhs, claims.user_token));
        self.current_fingerprint = fingerprint;
        self.current_client = Some(next_client.clone());
        Some(next_client)
    }
}
